server/recruitment.py

```diff
@@ -202,55 +246,107 @@ def get_questions(current_user=None):
+        raise ForbiddenException(f'User {user_id} is not an applicant')
+    application = Application.get_for_user(user_id)
+    if not application:
+        application = Application(user_id=user_id)
+    for answer in answers:
+        answer_record = Answer.query\
+            .filter_by(question_id=answer['question_id'], application_id=application.id)\
+            .one_or_none()
+        if not answer_record:
+            answer_record = Answer(question_id=answer['question_id'], application_id=application.id)
+            application.answers.append(answer_record)
+        answer_record.text = answer['text']
+    db.session.commit()
 
 
 def get_answers(user_id, current_user=None):
     if not db.session.query(db.exists().where(User.id==user_id)).scalar():
         raise BadRequestException('User with id={} does not exist.'.format(user_id))
     user = User.get(user_id)
+    current_user = current_user or user
     if not has_applicant_access(current_user, user, self_access=True):
         raise ForbiddenException('User {} does not have access to user {}'.format(current_user, user_id))
 
-    application_id = get_user_application(user_id).id
-    # get a dict keyed by question id of questions & answers
-    response = {}
+    application = get_user_application(user_id)
+    if not application:
+        raise BadRequestException('User with id={} has no application.'.format(user_id))
     questions = get_questions()
-    answer_query = db.session.query(Answer.question_id, Answer.text).filter(Answer.application_id==application_id)
-    answers = {item[0]: item[1] for item in answer_query}
-    for question_id in questions:
-        answer = answers[question_id] if question_id in answers else ""
-        response[question_id] = {
-            "question": questions[question_id],
-            "user_id": user_id,
-            "answer": answer,
-        }
+    response = {'questions':{}, 'has_application': False}
+    if application:
+        response['has_application'] = True
+        application_id = get_user_application(user_id).id
+        # get a dict keyed by question id of questions & answers
+        answer_query = db.session.query(Answer.question_id, Answer.text)\
+            .filter(Answer.application_id == application_id)
+        answers = {item[0]: item[1] for item in answer_query}
+        for question_id in questions:
+            answer = answers[question_id] if question_id in answers else ""
+            response['questions'][question_id] = {
+                'question': questions[question_id],
+                'user_id': user_id,
+                'answer': answer,
+            }
+    else:
+        # no application yet, create empty answers
+        for question_id in questions:
+            response['questions'][question_id] = {
+                'question': questions[question_id],
+                'user_id': user_id,
+                'answer': '',
+            }
     return response
 
 
+def start_application(current_user=None):
```

Comment:
We will need to actually call this from somewhere, let's talk about where in chat. Main possibilities are as an endpoint which means there's a "create application" button somewhere for new users, or automatically when a user provides scopes (which also requires we delete said application when an admin upgrades an account to recruiter).

Reply:
yes did that too :)

/api/recruits/start_application